pkg/kube/activity.go

```diff
@@ -407,6 +407,22 @@ func (k *PromoteStepActivityKey) GetOrCreatePromoteUpdate(activities typev1.Pipe
 	return a, s, p, p.Update, created, err
 }
 
+// Almost making a full Update, but using Patch to work around problem with sometimes getting http status 409 back
```

Comment:
So this is not really fixing the race condition which the K8s API server helpfully alerted you to with the 409 error. It is just masking it, by clobbering any other changes to the `spec` that happened to come in from another client. To my mind that is worse. At least without this patch, you can sometimes recover _and_ behave correctly simply by retrying the entire operation.

The real fix would be to review every use of a `PUT` call, which seems to be wrapped in a generated `Update` method, and try to replace it with a `PATCH`. [Here](https://github.com/jenkins-x/jx/blob/d2b2e75f1764b76c1a769e223f6c0c3601744e9a/pkg/jx/cmd/step_collect.go#L248:L259) is a simple example.

The trouble is that the generated `Patch` methods, as opposed to `Update` methods, lack type safety: you need to [manually construct a JSON patch body](https://github.com/jenkins-x/jx/blob/d2b2e75f1764b76c1a769e223f6c0c3601744e9a/pkg/kube/serviceaccount/serviceaccounts.go#L15:L22) before [applying it](https://github.com/jenkins-x/jx/blob/d2b2e75f1764b76c1a769e223f6c0c3601744e9a/pkg/kube/serviceaccount/serviceaccounts.go#L38:L50), which is a lot of work and looks error-prone. In fact the example I linked to seems to be the only such case in this whole repo so far! Compare the manually defined JSON structure

```go
type JsonPatch struct {
	ImagePullSecret *[]ImagePullSecret `json:"imagePullSecrets"`
}
type ImagePullSecret struct {
	Name string `json:"name"`
}
```

to the definition of the struct field in the official API class:

```go
ImagePullSecrets []LocalObjectReference `json:"imagePullSecrets,omitempty" protobuf:"bytes,3,rep,name=imagePullSecrets"`
// …
type LocalObjectReference struct {
	Name string `json:"name,omitempty" protobuf:"bytes,1,opt,name=name"`
}
```

Reply:
I suppose a less cumbersome approach would be to make `Update` methods take an old and new object, to be called like this:

```diff
diff --git a/pkg/jx/cmd/step_collect.go b/pkg/jx/cmd/step_collect.go
index 9d7a90f5..96d020de 100644
--- a/pkg/jx/cmd/step_collect.go
+++ b/pkg/jx/cmd/step_collect.go
@@ -249,11 +249,12 @@ func (o *StepCollectOptions) collectGitURL(sourceURL string) (err error) {
 		if err != nil {
 			return err
 		}
+		oldA := *a
 		a.Spec.Attachments = append(a.Spec.Attachments, jenkinsv1.Attachment{
 			Name: classifier,
 			URLs: urls,
 		})
-		_, err = client.JenkinsV1().PipelineActivities(ns).Update(a)
+		_, err = client.JenkinsV1().PipelineActivities(ns).Update(&oldA, a)
 		if err != nil {
 			return err
 		}
```

The implementation could optimistically send a `PUT` request, assuming the API server can be relied on to check `resourceVersion`s, and if it gets a 409 then use reflection to compute the minimal diff between the two trees and generate a `PATCH` request.

Alternately, we could just avoid `PATCH` alttogether and retry the whole block that failed in a `PUT`:

```diff
diff --git a/pkg/jx/cmd/step_collect.go b/pkg/jx/cmd/step_collect.go
index 9d7a90f5..48b11c69 100644
--- a/pkg/jx/cmd/step_collect.go
+++ b/pkg/jx/cmd/step_collect.go
@@ -245,6 +245,7 @@ func (o *StepCollectOptions) collectGitURL(sourceURL string) (err error) {
 				Build:    build,
 			},
 		}
+		for {
 			a, _, err := key.GetOrCreate(activities)
 			if err != nil {
 				return err
@@ -254,10 +255,15 @@ func (o *StepCollectOptions) collectGitURL(sourceURL string) (err error) {
 				URLs: urls,
 			})
 			_, err = client.JenkinsV1().PipelineActivities(ns).Update(a)
-		if err != nil {
+			if err == nil {
+				break
+			} else if strings.Contains(err.Error(), "409") {
+				continue
+			} else {
 				return err
 			}
 		}
+	}
 	return nil
 }
```

though I assume there is some more type-safe way of accessing the error code. (Seems there is a `StatusError` and you can check for `StatusReasonConflict` or something.)

Reply:
I think it's much better to, as you say, clobber the CRD than fail. Especially since at times  promotions has failed repeatedly on 409.

I did start a fix that did minimal patches, but I realised that the fix would become unreasonably big and complex.